Add docs on the main function, argv and return codes (#470)
spylang/spy: docs/src/howto/main_function.md

title: The Main Function
---

All SPy programs must have a `main` function. The main function is the entry point of the program, and it is where the execution of the program begins.

```py
def main() -> None:
    print("Hello world")
```

> Not every `.spy` module needs a main function, but the module invoked by, e.g. `spy execute foo.spy` must contain a main function.

Reviewer: ```suggestion
Not every `.spy` module needs a main function, but the module entry point invoked by, e.g. `spy foo.spy` or `spy build foo.spy` must contain a main function.
```

Reviewer: I think this is a big more correct, but it become more nuanced, because e.g. `spy build --target lib` and `spy build --target py-cffi` do not require a `main`.

Maybe we can reword the section by starting this way:

> The execution of a SPy program starts from a `main` function; you need a `main` if you try to run the interpreter (`spy foo.spy`) or to build an executable (`spy build foo.spy`). You don't need a `main` if you build a library (e.g. `spy build --target lib foo.spy`).

However, I don't fully like the wording. What do you think?

## Return Codes

The `main` function may be typed to return an `int` (`i32`). If so, the return value of the main function will be the return value of the program:

```py
#retcode.spy
def main() -> int
    return 123
```
```
$ uv run spy retcode.spy
$ echo $?
123
```

> ## Passed Arguments (Interpretter Only)

Reviewer 2: ```suggestion
## Passed Arguments (Interpreter Only)
```

Reviewer: as @kanin-kearpimy already pointed out, the arguments works also on compiled mode now 🎉.

Maybe instead of "passed arguments" we can call this section "Accessing command line arguments"?

> If the `main` function accepts a list of strings as an argument, the SPy program will accept arguments from the command line, both when running in interpretted

Reviewer 2: ```suggestion
If the `main` function accepts a list of strings as an argument, the SPy program will accept arguments from the command line, both when running in interpreted
```

Seems the sentence is not finished?

```py
#args.spy

> def main(args: list[str]) -> None:

Reviewer: I think we should use `argv` as the canonical name, for consistency to CPython

    print(args[1])
```
```
$ uv run spy args.spy 999
999
```

> As with CPython, args[0] is the name of the string passed to the uv runtime:

Reviewer: probably we should say that this is the equivalent of CPython's `sys.argv`
